File: native/android/session/src/main/kotlin/com/vocare/translate/session/offline/OfflineSessionViewModel.kt
package com.vocare.translate.session.offline

import androidx.lifecycle.ViewModel
import com.vocare.translate.core.model.Language
import com.vocare.translate.core.model.LiveState
import com.vocare.translate.core.model.SessionConfig
import com.vocare.translate.core.model.SessionPhase
import com.vocare.translate.core.model.Side
import com.vocare.translate.core.model.SideState
import com.vocare.translate.core.model.SpeechLocales
import com.vocare.translate.core.model.Turn
import com.vocare.translate.core.model.TurnView
import com.vocare.translate.core.offline.OfflineEngine
import com.vocare.translate.core.offline.PairStatus
import com.vocare.translate.core.offline.Recognition
import com.vocare.translate.session.EndReason
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Hold → on-device recognition, release → on-device translation → on-device
 * synthesis (CONTRACT §4, JSX `OfflineLiveScreen`). Never touches the network
 * and, like the JSX, does not write history.
 */
class OfflineSessionViewModel(
    private val config: SessionConfig,
    private val engine: OfflineEngine,
    dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
) : ViewModel() {

    private enum class Phase { IDLE, LISTENING, TRANSLATING, SPEAKING }

    /** One hold/release operation. Only the current one may touch state (JSX `operationRef`). */
    private class Operation(val side: Side) {
        var released = false
        var recognition: Recognition? = null
        val started = Job()
    }

    private val scope = CoroutineScope(SupervisorJob() + dispatcher)
    private val turns = mutableListOf<Turn>()
    private var holding: Side? = null
    private var phase = Phase.IDLE
    private var sessionPhase: SessionPhase = SessionPhase.Live
    private var notice: String? = null
    private var partial = ""
    private var elapsed = 0
    private var operation: Operation? = null
    private var releaseJob: Job? = null
    private val timer: Job = scope.launch {
        while (isActive) {
            delay(1_000)
            elapsed += 1
            publish()
        }
    }

    private val _state = MutableStateFlow(render())
    val state: StateFlow<LiveState> = _state.asStateFlow()

    private val _endReason = MutableStateFlow<EndReason?>(null)
    val endReason: StateFlow<EndReason?> = _endReason.asStateFlow()

    val transcript: List<Turn> get() = turns.toList()

    fun hold(side: Side) {
        if (operation != null || phase != Phase.IDLE || sessionPhase is SessionPhase.Ended) return
        val op = Operation(side)
        operation = op
        notice = null
        partial = ""
        holding = side
        phase = Phase.LISTENING
        publish()

        val from = lang(side)
        scope.launch {
            try {
                when (engine.pairStatus(config.langA, config.langB)) {
                    PairStatus.INSTALLED -> Unit
                    // Installable in-app; a declined download is not an error, just no turn.
                    PairStatus.SUPPORTED -> if (!engine.prepare(config.langA, config.langB)) {
                        throw OfflineSpeechException("Download both translation languages before starting an offline session.")
                    }
                    PairStatus.UNSUPPORTED -> throw OfflineSpeechException(
                        "${Language.of(config.langA).label} and ${Language.of(config.langB).label} cannot be translated on this device.",
                    )
                }
                if (operation !== op) return@launch
                val recognition = engine.startRecognition(SpeechLocales.recognition(from)) { text ->
                    if (operation === op && holding == side) {
                        partial = text
                        publish()
                    }
                }
                if (operation !== op) {
                    recognition.cancel()
                    return@launch
                }
                op.recognition = recognition
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                if (operation === op) {
                    operation = null
                    holding = null
                    phase = Phase.IDLE
                    notice = e.message ?: "Could not start on-device speech recognition."
                    publish()
                }
            } finally {
                op.started.complete()
            }
        }
    }

    fun release(side: Side) {
        val op = operation ?: return
        if (op.side != side || op.released) return
        op.released = true
        holding = null
        phase = Phase.TRANSLATING
        publish()
        releaseJob = scope.launch { finishTurn(op) }
    }

    private suspend fun finishTurn(op: Operation) {
        val from = lang(op.side)
        val to = lang(op.side.other)

        op.started.join()
        if (operation !== op) return
        // Start failed: hold() already reported it and reset the phase.
        val recognition = op.recognition ?: return

        val heard = try {
            recognition.stop().trim()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            settle(op, e.message ?: "Could not finish speech recognition.")
            return
        }
        if (operation !== op) return
        if (heard.isEmpty()) {
            settle(
                op,
                "Nothing was recognised in ${Language.of(from).label}. " +
                    "On-device recognition for ${Language.of(from).label} may not be installed — " +
                    "download the recognition language in your device speech settings.",
            )
            return
        }

        val translated = try {
            engine.translate(heard, from, to).trim()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            ""
        }
        if (operation !== op) return
        if (translated.isEmpty()) {
            settle(
                op,
                "Could not translate ${Language.of(from).label} to ${Language.of(to).label} on this device. " +
                    "The language may not be downloaded — check Settings.",
            )
            return
        }

        turns += Turn(
            speaker = op.side.name,
            speakerName = name(op.side),
            originalLang = from,
            translatedLang = to,
            original = heard,
            translated = translated,
            side = op.side.name,
            from = from,
            to = to,
        )
        partial = ""
        phase = Phase.SPEAKING
        publish()
        try {
            engine.speak(translated, SpeechLocales.speech(to))
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            if (operation === op) notice = e.message ?: "Could not speak the translation offline."
        }
        if (operation !== op) return
        operation = null
        phase = Phase.IDLE
        publish()
    }

    private fun settle(op: Operation, message: String) {
        if (operation !== op) return
        operation = null
        phase = Phase.IDLE
        partial = ""
        notice = message
        publish()
    }

    fun end() {
        if (sessionPhase is SessionPhase.Ended) return
        cancelOperation()
        timer.cancel()
        sessionPhase = SessionPhase.Ended
        _endReason.value = EndReason.USER
        publish()
    }

    private fun cancelOperation() {
        val op = operation ?: return
        operation = null
        releaseJob?.cancel()
        op.recognition?.let { recognition -> scope.launch { recognition.cancel() } }
        holding = null
        phase = Phase.IDLE
        partial = ""
    }

    override fun onCleared() {
        cancelOperation()
        scope.cancel()
    }

    // ── projection ───────────────────────────────────────────────────────────

    private fun publish() {
        _state.value = render()
    }

    private fun render(): LiveState = LiveState(
        sideA = renderSide(Side.A),
        sideB = renderSide(Side.B),
        elapsedSeconds = elapsed,
        phase = sessionPhase,
        badge = BADGE,
        notice = notice,
    )

    private fun renderSide(side: Side): SideState {
        val busy = phase != Phase.IDLE
        val mine = holding == side
        val turnViews = turns.mapIndexed { i, t -> TurnView(key = i, turn = t, mine = t.side == side.name) }
        return SideState(
            name = name(side),
            langCode = lang(side),
            langLabel = Language.of(lang(side)).label,
            pressing = mine,
            disabled = sessionPhase is SessionPhase.Ended || (busy && !mine),
            dimmed = busy && !mine,
            status = when {
                mine -> "Recording"
                phase == Phase.TRANSLATING -> "Translating..."
                phase == Phase.SPEAKING -> "Speaking..."
                holding != null -> "Listening to ${name(side.other)}"
                turns.isNotEmpty() -> "Ready - speak again"
                else -> "Ready"
            },
            turns = turnViews,
            // The partial belongs on the speaker's own half — it is what they are saying.
            note = if (mine && partial.isNotEmpty()) partial else null,
        )
    }

    private fun lang(side: Side) = if (side == Side.A) config.langA else config.langB
    private fun name(side: Side) = if (side == Side.A) config.nameA else config.nameB

    companion object {
        const val BADGE = "ON DEVICE"
    }
}

File: native/ios/VocaKit/Sources/VocaSession/Offline/OfflineSessionViewModel.swift
import Foundation
import VocaKit

// MARK: - Offline session (CONTRACT §4, mirrors JSX `OfflineLiveScreen`)
//
// Zero network. Each hold-and-release on side S runs STT → translate → TTS
// through the injected `OfflineEngine`, appending one turn. The same panel
// semantics as the cloud session: a turn spoken by A is "You said" on A's
// panel and "Translated" on B's.

@MainActor
public final class OfflineSessionViewModel: ObservableObject {
    @Published public private(set) var state: LiveState

    public static let badge = "On device"

    public enum Phase: Equatable, Sendable {
        case idle, listening, translating, speaking
    }

    public struct OfflineTurn: Equatable, Sendable {
        public let side: Side
        public let from: String
        public let to: String
        public let original: String
        public let translated: String
    }

    /// One hold-and-release. Identity-compared so a stale operation that wakes
    /// up after `end()` (or after an error reset) cannot touch state.
    @MainActor private final class Operation {
        let side: Side
        var released = false
        var start: Task<Void, Error>?
        init(side: Side) { self.side = side }
    }

    private let config: SessionConfig
    private let engine: OfflineEngine
    private let scheduler: SessionScheduler
    private let tasks = TaskTracker()

    private var langA: String
    private var langB: String
    private(set) public var turns: [OfflineTurn] = []
    private var holding: Side?
    private(set) public var phase: Phase = .idle
    private var error: String?
    private var partial = ""
    private var elapsedSeconds = 0
    private var ended = false
    private var operation: Operation?
    private var ticker: SchedulerToken?

    public init(config: SessionConfig, engine: OfflineEngine, scheduler: SessionScheduler = TaskScheduler()) {
        self.config = config
        self.engine = engine
        self.scheduler = scheduler
        self.langA = config.langA
        self.langB = config.langB
        self.state = LiveState(
            sideA: SideState(name: config.nameA, langCode: config.langA, langLabel: SessionLanguages.label(config.langA)),
            sideB: SideState(name: config.nameB, langCode: config.langB, langLabel: SessionLanguages.label(config.langB)),
            phase: .live, badge: OfflineSessionViewModel.badge
        )
        ticker = scheduler.repeating(every: 1) { [weak self] in
            guard let self, !self.ended else { return }
            self.elapsedSeconds += 1
            self.publish()
        }
        publish()
    }

    /// Await in-flight hold/release work (tests).
    public func settle() async { await tasks.settle() }

    // MARK: Push-to-talk

    public func hold(side: Side) {
        guard !ended, operation == nil, phase == .idle else { return }
        let op = Operation(side: side)
        operation = op
        error = nil
        partial = ""
        holding = side
        phase = .listening
        publish()

        let from = side == .a ? langA : langB
        let a = langA, b = langB
        let engine = self.engine
        op.start = Task { @MainActor [weak self] in
            // §4: packs are installable in-app. A `supported` pair is
            // downloaded here (Apple's own sheet asks the user); a declined
            // download is a notice, not a hard failure.
            switch await engine.pairStatus(a, b) {
            case .installed:
                break
            case .supported:
                guard try await engine.prepare(a, b) else {
                    throw OfflineEngineError.translationUnavailable(from: a, to: b)
                }
            case .unsupported:
                throw OfflineEngineError.translationUnavailable(from: a, to: b)
            }
            guard let self, self.operation === op else { throw OfflineEngineError.cancelled }
            try await engine.startRecognition(locale: SessionLanguages.speechLocale(from)) { text in
                Task { @MainActor [weak self] in
                    guard let self, self.operation === op, self.holding == side else { return }
                    self.partial = text
                    self.publish()
                }
            }
        }
        tasks.track { [weak self] in
            guard let self else { return }
            do {
                try await op.start?.value
            } catch {
                guard self.operation === op else { return }
                self.operation = nil
                self.holding = nil
                self.phase = .idle
                self.error = OfflineSessionViewModel.message(for: error, from: from, to: side == .a ? b : a,
                                                             fallback: "Could not start on-device speech recognition.")
                self.publish()
            }
        }
    }

    public func release(side: Side) {
        guard let op = operation, op.side == side, !op.released else { return }
        op.released = true
        let from = side == .a ? langA : langB
        let to = side == .a ? langB : langA
        holding = nil
        phase = .translating
        publish()

        tasks.track { [weak self] in
            guard let self else { return }

            // 1. Finish recognition.
            var heard = ""
            do {
                try await op.start?.value
                guard self.operation === op else { return }
                heard = try await self.engine.finishRecognition()
                guard self.operation === op else { return }
            } catch {
                guard self.operation === op else { return }
                self.finishOperation(error: OfflineSessionViewModel.message(
                    for: error, from: from, to: to, fallback: "Could not finish speech recognition."))
                return
            }
            heard = heard.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !heard.isEmpty else {
                self.finishOperation(error:
                    "Nothing was recognised in \(SessionLanguages.label(from)). "
                    + "On-device recognition for \(SessionLanguages.label(from)) may not be installed — "
                    + "download the recognition language in your device speech settings.")
                return
            }

            // 2. Translate.
            self.phase = .translating
            self.publish()
            let translated = (try? await self.engine.translate(heard, from: from, to: to))?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard self.operation === op else { return }
            guard !translated.isEmpty else {
                self.finishOperation(error:
                    "Could not translate \(SessionLanguages.label(from)) to \(SessionLanguages.label(to)) on this device. "
                    + "The language may not be downloaded — check Settings.")
                return
            }

            self.turns.append(OfflineTurn(side: side, from: from, to: to, original: heard, translated: translated))
            self.partial = ""
            self.phase = .speaking
            self.publish()

            // 3. Speak with a local voice.
            do {
                try await self.engine.speak(translated, locale: SessionLanguages.speechLocale(to))
            } catch {
                if self.operation === op {
                    self.error = OfflineSessionViewModel.message(
                        for: error, from: from, to: to, fallback: "Could not speak the translation offline.")
                }
            }
            guard self.operation === op else { return }
            self.finishOperation(error: self.error)
        }
    }

    private func finishOperation(error: String?) {
        operation = nil
        holding = nil
        phase = .idle
        partial = ""
        self.error = error
        publish()
    }

    /// Swap A/B languages between turns. Earlier bubbles were authored under
    /// the old assignment, so they are cleared (JSX `swapLanguages`).
    public func swapLanguages() {
        guard phase == .idle, operation == nil else { return }
        swap(&langA, &langB)
        turns.removeAll()
        partial = ""
        error = nil
        publish()
    }

    public func end() {
        guard !ended else { return }
        ended = true
        ticker?.cancel(); ticker = nil
        let op = operation
        operation = nil
        holding = nil
        phase = .idle
        let engine = self.engine
        if op != nil {
            tasks.track { await engine.cancelRecognition() }
        }
        state.phase = .ended(.user)
    }

    // MARK: Projection

    private static func message(for error: Error, from: String, to: String, fallback: String) -> String {
        switch error as? OfflineEngineError {
        case .permissionDenied:
            return "Speech recognition permission was declined. Enable microphone access for Voca in your device settings."
        case .recognitionUnavailable:
            return "On-device speech recognition is not available on this device."
        case .recognitionTimedOut:
            return "Speech recognition timed out. Please try again."
        case .translationUnavailable:
            return "Download both translation languages before starting an offline session."
        case .noLocalVoice:
            return "No offline voice is installed for this language. Download a voice in your device text-to-speech settings."
        case .recognitionFailed(let text), .translationFailed(let text):
            return text
        case .cancelled, .none:
            return fallback
        }
    }

    private func publish() {
        guard !ended else { return }
        let busy = phase != .idle

        func side(_ side: Side) -> SideState {
            let name = side == .a ? config.nameA : config.nameB
            let otherName = side == .a ? config.nameB : config.nameA
            let code = side == .a ? langA : langB
            let status: String
            if holding == side { status = "Recording" }
            else if phase == .translating { status = "Translating..." }
            else if phase == .speaking { status = "Speaking..." }
            else if holding != nil { status = "Listening to \(otherName)" }
            else { status = turns.isEmpty ? "Ready" : "Ready - speak again" }

            let views = turns.enumerated().map { index, turn in
                TurnView(id: index, original: turn.original, translated: turn.translated, mine: turn.side == side)
            }
            return SideState(
                name: name, langCode: code, langLabel: SessionLanguages.label(code),
                pressing: holding == side,
                disabled: busy && holding != side,
                dimmed: busy && holding != side,
                status: status,
                turns: views,
                note: (holding == side && !partial.isEmpty) ? partial : nil
            )
        }

        state = LiveState(
            sideA: side(.a), sideB: side(.b),
            elapsed: TimeInterval(elapsedSeconds),
            phase: .live,
            badge: OfflineSessionViewModel.badge,
            notice: error
        )
    }
}
